src/rmt.rs

```diff
@@ -0,0 +1,332 @@
+        self.config.__bindgen_anon_1.tx_config.idle_output_en = enabled;
+        self
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct PulseTicks(u16);
+
+impl PulseTicks {
+    const MAX: u16 = 32767;
+
+    /// PulseTick needs to be unsigned 15 bits: 0-32767 inclusive.
+    ///
+    /// It will return `None` if it's out of range.
+    pub fn new(v: u16) -> Option<Self> {
+        if v > 32767 {
+            // TODO: Not sure of a nicer way to deal with this? Allow an overflow like rust does?
+            None
+        } else {
+            Some(Self(v))
+        }
+    }
+
+    /// Use the maximum value of 32767.
+    pub fn max() -> Self {
+        Self(Self::MAX)
+    }
+}
+
+pub struct Writer {
+    config: rmt_config_t,
+
+    // This must be ManuallyDrop to ensure that it isn't automatically dropped before the driver is
+    // done using the items.
+    // TODO: Manually drop this!
+    items: ManuallyDrop<Vec<rmt_item32_t>>,
+
+    // An item that has had only its first half populated.
+    half_inserted: Option<rmt_item32_t>,
+
+    ticks_per_ns: Option<f32>,
+}
+
+impl Writer {
+    pub fn new(config: WriterConfig) -> Result<Writer, EspError> {
+        let s = Writer {
+            config: config.config,
+            items: Default::default(),
+            half_inserted: None,
+            ticks_per_ns: None,
+        };
+
+        unsafe {
+            esp!(rmt_config(&s.config))?;
+            esp!(rmt_driver_install(s.config.channel as u32, 0, 0))?;
+        }
+
+        Ok(s)
+    }
+
+    pub fn counter_clock(&self) -> Result<u32, EspError> {
+        let mut ticks_hz: u32 = 0;
+        esp!(unsafe { rmt_get_counter_clock(self.config.channel, &mut ticks_hz) })?;
+        Ok(ticks_hz)
+    }
+
+    /// Creates a Pulse based on Nanoseconds.
+    ///
+    /// Returns None if the resulting ticks from the ns conversion is too high.
+    /// See `PulseTicks` for details.
+    ///
+    /// This function exists on `Writer` because of the internal call to `rmt_get_counter_clock`
+    /// requires a channel which is managed by `Writer`.
+    pub fn pulse_ns(&mut self, pin_state: PinState, ns: u32) -> Result<Option<Pulse>, EspError> {
+        let ticks_per_ns = match &self.ticks_per_ns {
+            None => {
+                let ticks_hz = self.counter_clock()?;
+                let ticks_per_ns = ticks_hz as f32 / 1e9;
+                self.ticks_per_ns = Some(ticks_per_ns);
+                self.ticks_per_ns.as_ref().unwrap()
+            }
+            Some(t) => t,
+        };
+        let ticks = (ticks_per_ns * ns as f32) as u16;
+        match PulseTicks::new(ticks) {
+            Some(ticks) => Ok(Some(Pulse::new(pin_state, ticks))),
+            None => Ok(None),
+        }
+    }
+
+    pub fn add<I>(&mut self, pulses: I) -> Result<(), EspError>
+    where
+        I: IntoIterator<Item = Pulse>,
+    {
+        for pulse in pulses {
+            println!("{:?} {}", &pulse.pin_state, pulse.ticks.0);
```

Comment:
Remove the `println!` in the final version. The driver should be `no_std` compatible (so no `println!`), and having debugging `println!` statements around of course not ideal.

Reply:
Sorry, this was some debugging :lady_beetle: